src/join/core.jl

```diff
@@ -428,84 +509,306 @@ function _innerjoin_postprocess_int(left::AbstractVector{<:Union{Integer, Missin
+            return _innerjoin_unsorted_int(left_col, right_col)
+        else
+            return _innerjoin_unsorted(left_col, right_col)
         end
     else
-        for i in eachindex(left_cols, right_cols)
-            rc = right_cols[i]
-            lc = left_cols[i]
+        if left_col isa AbstractVector{<:Union{Integer, Missing}} &&
+           right_col isa AbstractVector{<:Union{Integer, Missing}}
+            return reverse(_innerjoin_unsorted_int(right_col, left_col))
+        else
+            return reverse(_innerjoin_unsorted(right_col, left_col))
+        end
+    end
 
-            rc_refs = DataAPI.refarray(rc)
-            rc_refpool = DataAPI.refpool(rc)
-            rc_invrefpool = DataAPI.invrefpool(rc)
-            if check_mapping_allowed(lc, rc_refs, rc_refpool, rc_invrefpool)
-                right_cols[i] = rc_refs
-                left_cols[i] = map2refs(lc, rc_invrefpool)
+    error("unreachable reached")
+end
+
+### semijoin logic
+
+function _semijoin_sorted(left::AbstractArray, right::AbstractArray, seen_rows::AbstractVector{Bool})
+    left_n = length(left)
+    right_n = length(right)
+
+    @assert left_n > 0 && right_n > 0
+
+    left_cur = 1
+    left_val = left[left_cur]
+    left_new, left_tmp = find_next_range(left, left_cur, left_val)
+
+    right_cur = 1
+    right_val = right[right_cur]
+    right_new, right_tmp = find_next_range(right, right_cur, right_val)
+
+    while left_cur <= left_n && right_cur <= right_n
+        if isequal(left_val, right_val)
+            seen_rows[left_cur:left_new - 1] .= true
+            left_cur, left_val = left_new, left_tmp
+            left_new, left_tmp = find_next_range(left, left_cur, left_val)
+            right_cur, right_val = right_new, right_tmp
+            right_new, right_tmp = find_next_range(right, right_cur, right_val)
+        elseif isless(left_val, right_val)
+            left_cur, left_val = left_new, left_tmp
+            left_new, left_tmp = find_next_range(left, left_cur, left_val)
+        else
+            right_cur, right_val = right_new, right_tmp
+            right_new, right_tmp = find_next_range(right, right_cur, right_val)
+        end
+    end
+
+    return seen_rows
+end
+
+# optimistically assume that shorter table does not have duplicates in on column
+# if this is not the case we call _semijoin_dup
+# which efficiently uses the work already done and continues with the more
+# memory expensive algorithm that allows for duplicates
+function _semijoin_unsorted(left::AbstractArray, right::AbstractArray{T},
+                            seen_rows::AbstractVector{Bool},
+                            ::Val{right_shorter}) where {T, right_shorter}
+    if right_shorter
+        @assert length(left) == length(seen_rows)
+    else
+        @assert length(right) == length(seen_rows)
+    end
+
+    dict = Dict{T, Int}()
+
+    right_len = length(right)
+    sizehint!(dict, 2 * min(right_len, typemax(Int) >> 2))
```

Comment:
Can you add a comment explaining how the size is chosen? Why the × 2 in particular?

Reply:
I have added a comment